File: Cursos e Vídeos/Python/Corey Schafer (Youtube)/Programming Terms/02 First-Class Functions.py
# -*- coding: utf-8 -*-
"""
Created on Tue Dec 29 12:09:37 2020

@author: pedro
"""

def square(x):
    return x * x

def cube(x):
    return x * x * x


f = square(5)


print(square)
print(f)

f = square

print(square)
print(f(5))

##################

def my_map(func,arg_list):
    result = []
    for i in arg_list:
        result.append(func(i))
    return result

squares = my_map(square, [1,2,3,4,5])

squares = my_map(cube, [1,2,3,4,5])

print(squares)

##################

def logger(msg):
    
    def log_message():
        print('Log: ', msg)
        
    return log_message

log_hi = logger('Hi')
log_hi()

##################

def html_tag(tag):
    
    def wrap_text(msg):
        print('<{0}>{1}</{0}>'.format(tag,msg))
    
    return wrap_text

print_h1 = html_tag('h1')
print_h1('Test Headline!')
print_h1('Another Headline!')

print_p = html_tag('p')
print_p('Test Paragraph!')

File: JavaScript/Cursos e Vídeos/Corey Schafer (Youtube)/Programming Terms/02 First-Class Functions.js

/* function square(x){
    return x * x;
}

var f = square(5)

console.log(square)
console.log(f) */

// -----------------------

/* 
function square(x){
    return x * x;
}

var f = square

console.log(square)
console.log(f(5))

function my_map(func, arg_list){
    result = [];
    for(var i=1; i<arg_list.length; i++){
        result.push(func(i))
    }
    return result;
}

var squares = my_map(square, [1,2,3,4,5])
console.log(squares)

var squares = my_map(cube, [1,2,3,4,5])
console.log(squares)


function cube(x){
    return x * x * x
} */

// -----------------------

/* function logger(msg) {
    function log_message() {
        console.log('Log: ' + msg)
    }
    return log_message
}

log_hi = logger('Hi')
log_hi() */

// -----------------------

function html_tag(tag){
    function wrap_text(msg){
        console.log('<' + tag + '>' + msg + '</' + tag + '>')
    }
    return wrap_text
}

print_h1 = html_tag('h1')
print_h1('Test Headline!')
print_h1('Another Headline!')

print_p = html_tag('p')
print_p('Test Paragraph')





















